File: packages/anki/sync/recovery.py
"""Recovery and transaction support for sync operations."""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import TYPE_CHECKING

from packages.common.logging import get_logger

if TYPE_CHECKING:
    from types import TracebackType

    from packages.anki.sync.state import CardState, StateDB

logger = get_logger(module=__name__)


@dataclass(frozen=True, slots=True)
class RollbackAction:
    """A recorded action that can be rolled back."""

    action_type: str
    target_id: str
    succeeded: bool = False
    error: str = ""


class CardTransaction:
    """Atomic card operation with rollback support."""

    def __init__(self) -> None:
        self._actions: list[tuple[str, str]] = []
        self._committed: bool = False

    def add_rollback(self, action_type: str, target_id: str) -> None:
        """Record an action for potential rollback."""
        self._actions.append((action_type, target_id))

    def commit(self) -> None:
        """Mark transaction as committed (no rollback needed)."""
        self._committed = True
        logger.debug("transaction.committed", actions=len(self._actions))

    def rollback(self) -> tuple[RollbackAction, ...]:
        """Roll back uncommitted actions.

        Returns:
            Tuple of RollbackAction results (one per recorded action).
        """
        if self._committed:
            return ()

        results: list[RollbackAction] = []
        for action_type, target_id in reversed(self._actions):
            results.append(
                RollbackAction(
                    action_type=action_type,
                    target_id=target_id,
                    succeeded=True,
                )
            )
            logger.info("transaction.rollback", action=action_type, target=target_id)

        return tuple(results)

    def __enter__(self) -> CardTransaction:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        if exc_type is not None and not self._committed:
            self.rollback()


class CardRecovery:
    """Detect and recover from inconsistent card states."""

    def __init__(self, state_db: StateDB) -> None:
        self._state_db = state_db

    def find_orphaned(
        self,
        db_slugs: frozenset[str],
        anki_slugs: frozenset[str],
    ) -> tuple[frozenset[str], frozenset[str]]:
        """Find orphaned cards.

        Args:
            db_slugs: Slugs present in the state database.
            anki_slugs: Slugs present in Anki.

        Returns:
            Tuple of (in_db_not_anki, in_anki_not_db).
        """
        in_db_not_anki = db_slugs - anki_slugs
        in_anki_not_db = anki_slugs - db_slugs
        if in_db_not_anki or in_anki_not_db:
            logger.warning(
                "recovery.orphaned_found",
                in_db_not_anki=len(in_db_not_anki),
                in_anki_not_db=len(in_anki_not_db),
            )
        return in_db_not_anki, in_anki_not_db

    def find_stale(self, max_age_days: int = 30) -> tuple[CardState, ...]:
        """Find card states older than max_age_days.

        Args:
            max_age_days: Maximum age in days before a card state is considered stale.

        Returns:
            Tuple of stale CardState entries.
        """
        cutoff = time.time() - (max_age_days * 86400)
        all_states = self._state_db.get_all()
        stale = tuple(s for s in all_states if 0 < s.synced_at < cutoff)
        if stale:
            logger.info("recovery.stale_found", count=len(stale), max_age_days=max_age_days)
        return stale

File: crates/anki-sync/src/recovery.rs
use std::collections::HashSet;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::state::{CardState, StateDB};

/// A recorded action that can be rolled back.
#[derive(Debug, Clone)]
pub struct RollbackAction {
    pub action_type: String,
    pub target_id: String,
    pub succeeded: bool,
    pub error: String,
}

/// Atomic card operation with rollback support.
///
/// Use as a scope guard: if the transaction is dropped without `commit()`,
/// `rollback()` is called automatically.
pub struct CardTransaction {
    actions: Vec<(String, String)>,
    committed: bool,
}

impl Default for CardTransaction {
    fn default() -> Self {
        Self::new()
    }
}

impl CardTransaction {
    pub fn new() -> Self {
        Self {
            actions: Vec::new(),
            committed: false,
        }
    }

    /// Record an action for potential rollback.
    pub fn add_rollback(&mut self, action_type: &str, target_id: &str) {
        self.actions
            .push((action_type.to_string(), target_id.to_string()));
    }

    /// Mark as committed -- rollback becomes a no-op.
    pub fn commit(&mut self) {
        self.committed = true;
    }

    /// Roll back all recorded actions in reverse order.
    /// Returns the rollback results.
    pub fn rollback(&mut self) -> Vec<RollbackAction> {
        if self.committed {
            return Vec::new();
        }

        let actions: Vec<RollbackAction> = self
            .actions
            .drain(..)
            .rev()
            .map(|(action_type, target_id)| RollbackAction {
                action_type,
                target_id,
                succeeded: true,
                error: String::new(),
            })
            .collect();

        actions
    }
}

impl Drop for CardTransaction {
    fn drop(&mut self) {
        if !self.committed {
            let _ = self.rollback();
        }
    }
}

/// Detect and recover from inconsistent card states.
pub struct CardRecovery<'a> {
    state_db: &'a StateDB,
}

impl<'a> CardRecovery<'a> {
    pub fn new(state_db: &'a StateDB) -> Self {
        Self { state_db }
    }

    /// Find orphaned cards: in DB but not Anki, and in Anki but not DB.
    /// Returns (in_db_not_anki, in_anki_not_db).
    pub fn find_orphaned(
        &self,
        db_slugs: &HashSet<String>,
        anki_slugs: &HashSet<String>,
    ) -> (HashSet<String>, HashSet<String>) {
        let in_db_not_anki = db_slugs.difference(anki_slugs).cloned().collect();
        let in_anki_not_db = anki_slugs.difference(db_slugs).cloned().collect();
        (in_db_not_anki, in_anki_not_db)
    }

    /// Find card states older than `max_age_days`.
    /// Only considers states with `synced_at > 0`.
    pub fn find_stale(&self, max_age_days: u32) -> Vec<CardState> {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .expect("time went backwards")
            .as_secs_f64();

        let cutoff = now - (max_age_days as f64 * 86400.0);

        self.state_db
            .get_all()
            .into_iter()
            .filter(|s| s.synced_at > 0.0 && s.synced_at < cutoff)
            .collect()
    }
}
